### ISSUE TITLE
Scoped association tokens

### ISSUE THREAD

**Issue by jcnelson** [blockstack/gaia#142]
A question I get a lot is how to allow one user to write to a *limited subset* of another user's Gaia hub.  The way this would be achieved would be to expand Gaia's `associationToken` to include scopes.  I'm thinking we'd add the following:

* Scope writes to a particular prefix within the Gaia hub
* Allow writes but not list-files
* Allow list-files but not writes

Thoughts? @kantai

**Comment by markmhx** [comment]
This sounds intriguing. I'd be curious to hear about the functional use cases behind the question to better understand the types of data developers would allow users to write to each others' hubs.

**Comment by hstove** [comment]
@markmhx yes as Thomas said, this is exactly how I'd like to build out collaboration functionality in Kanstack.

I know it's maybe not a priority right now, but is there anything I can do to help move this along? I could definitely write some blockstack.js code to accommodate  this at some point.

**Comment by jcnelson** [comment]
I updated the issue to say "authentication token," since this is what users ask for.  They want multiple users to be able write to the same *bucket* in the same Gaia hub, not just write to the same Gaia hub.  To clarify the high-level difference:

* association tokens are what allow you to write to a particular Gaia hub
* authentication tokens are what allow you to write to a specific bucket in a particular Gaia hub

**Comment by markmhx** [comment]
Thanks for the clarifactions, @hstove and @jcnelson.

Can we also update the issue's title so that it's more imperative in nature, and as such describes to core task here. e.g. "Expand authentication token support to include scopes"?

If we already have authentication tokens but not these scopes, what are they currently used for? Allowing writes and list-files for all buckets in a given hub?

**Comment by kantai** [comment]
> If we already have authentication tokens but not these scopes, what are they currently used for? Allowing writes and list-files for all buckets in a given hub?

The authentication tokens are used to allow writes and list-files for a specific bucket. Scoping would allow us to create tokens which only grant write access to, say, specific paths.

**Comment by kantai** [comment]
Is it okay to not support list-files in the authentication token? I want to make sure that we're not encouraging a use case where applications are doing a lot of list-files queries when other strategies would work. If there's a compelling use case for non-owners to perform list-files, I'd be okay with that. But here's roughly what I'd like to see in a scoped auth token:

```
pathsAllowed: [ string ],
maxFileSize: number
```

What were you thinking for the pathsAllowed? Supporting wildcards, or just explicit paths?

**Comment by hstove** [comment]
Paths should definitely be a wildcard. My use-case would be to allow writing to files under a path like `/team/:id/*` (where `id` is explicit).

Also with my app developer hat on, list-files would be preferred, but I also would be fine with using other strategies.

**Comment by jcnelson** [comment]
@kantai The alternative to not implementing list-files is for each application that needs it implements their own list-files workalike (which Graphite and Stealthy have both done already).  Basically, whenever applications need collections-like functionality, they'll need list-files.  I suspect that this will apply to nearly every application where users share data.

**Comment by jcnelson** [comment]
@hstove Definitely agree that we should do wildcarding.  However, we'll want to be careful in the implementation -- we don't necessarily want to support full globs (this potentially opens a DDoS vector).

**Comment by hstove** [comment]
Yeah, from a developer experience, we should either implement an easy and standard way to list public files, or just implement a way for scoped auth tokens to do list-files.

**Comment by kantai** [comment]
> @kantai The alternative to not implementing list-files is for each application that needs it implements their own list-files workalike (which Graphite and Stealthy have both done already). Basically, whenever applications need collections-like functionality, they'll need list-files. I suspect that this will apply to nearly every application where users share data.

This is true of a lot of things, though. Apps can implement this without support from the Gaia hub, and do today. If this issue is one of interfaces, then the client library can do it. If the normal behavior of applications is using list-files on a user's gaia hub, then that hub itself is now responsible for serving read traffic, which is not preferable, and a major change in how users interact with the system.

**Comment by jcnelson** [comment]
> You mean not support things like /folder/**/*? I think that's perfectly fine, it can be only one folder deep.
Gaia doesn't have a notion of folders -- it's a key/value store.  Using the kleene star operator for describing matches is fine, but we should probably avoid character classes and match groups for now.

> Apps can implement this without support from the Gaia hub, and do today. If this issue is one of interfaces, then the client library can do it.

This is a good point.  If we end up not allowing list-files to be specified as a scope, then we should also commit to implementing this library as the de-facto standard way for doing data collections (so we don't end up with a bunch of incompatible mechanisms).

**Comment by jcnelson** [comment]
So, do we know which option we're talking?

Option 1:  allow list-files to be a scope, and in doing so, allow the creator of the list-files scope to restrict the user to paths that match specific wildcard(s)

Option 2:  do not allow list-files to be a scope.  Instead, offer developers a canonical spec and library for implementing what they need for listing files in another Gaia hub.

**Comment by kantai** [comment]
Yep -- I prefer going with Option 2 -- and we should just make the "scope key" an object, with a field that looks like "allowedPaths", which will allow us to add more scopes in a backwards-compatible way in the future. Adding a list-files risks really changing the way that developers interact with gaia hubs today, and I don't think we're ready to commit to that yet.

**Comment by hstove** [comment]
Yes I think option 2 is smart.

I don't fully understand how this would be implemented as a dev. Is this something I can call from blockstack.js after auth, or will it be part of the browser auth flow?

**Comment by jcnelson** [comment]
Thanks @kantai and @hstove for your input!

> I don't fully understand how this would be implemented as a dev. Is this something I can call from blockstack.js after auth, or will it be part of the browser auth flow?

What Stealthy and Graphite do is they have a canonical Gaia file called something like `collections.json`.  That file, in turn, lists all the names of the files belonging to that collection.  When the user wants to add or remove a file from the collection, the `collections.json` file would also be updated.  This means there would be one `getFile()` and two `putFile()` calls per update -- one `putFile()` to store the item itself, and one `putFile()` and one `getFIle()` to load, update, and store the associated `collections.json`.  I think @kantai was proposing that this process could be standardized and made into a library, so one application can read through another application's collections without having to know the specifics of how that app was built.

**Comment by hstove** [comment]
Sorry @jcnelson I didn't mean how to do collections, I meant how to request a scoped auth token.

**Comment by jcnelson** [comment]
My bad.

I don't think there's a method for requesting a scoped auth token yet.  We'd have to add it to `blockstack.js`.  Probably by modifying `makeV1GaiaAuthToken` to take a list of scopes and by making it public.

**Comment by hstove** [comment]
Awesome, yeah I knew it wasn't in the library yet, I just wasn't sure what that would look like. Thanks!

**Comment by jcnelson** [comment]
Okay, so I have an implementation that I'm almost ready to submit as a PR.  Just to be clear, the only change we're making is to allow the auth token signer to constrain where the bearer can write (i.e. the set of paths on which `putFile()` is allowed).  The change set does *not* touch `list-files` in any way.

My current code allows you to send an array of `AuthScopeType` objects, where each `AuthScopeType` contains a `scope` and a `domain` (both `string`).  Right now, there are to legal values for `scope`: `putFile` and `putFilePrefix`.

* The `putFile` scope lets you control the **exact** file to which the bearer can write.  For example, if the auth token contains the auth scope `{ scope: "putFile", domain: "team/123/info.json" }`, then the bearer can only call `putFile()` on `team/123/info.json`.

* The `putFilePrefix` scope lets you control the prefix string of the set of files to which a bearer can write.  For example, if the auth token contains the auth scope `{ scope: "putFilePrefix", domain: "team/123" }`, then the bearer can call `putFile()` on any file that starts with `team/123`, including `team/123456`, `team/123/info.json`, and so on (note that Gaia is a key-value store and assigns no semantic meaning to "/").

Is there anything else to this issue?  If not, I'll just submit my PR for your review.

**Comment by markmhx** [comment]
@jcnelson this is intended for @kantai to code review, correct?

**Comment by jcnelson** [comment]
@kantai at the very least, but probably also @hstove and yourself if you'd like to kick the tires (@hstove has a real-world use-case and I want to make sure this feature branch covers it).

**Comment by hstove** [comment]
I'll look at your code, but the spec you described earlier today works perfectly.

**Comment by kantai** [comment]
That soec sounds good to me @jcnelson --- fire away with the PR!

**Comment by markmhx** [comment]
I'll leave this to @hstove and @kantai since I don't want to be a blocker and think you're in good hands 🤗

Let me know if you do end up needing a third set of eyes though

**Comment by hstove** [comment]
Hey @jcnelson, wondering about this issue:

Can I de-authorize an auth token?

If I'm using this to provide functionality for teams, you'd need to handle the case where you remove a team member, and need to revoke their access. It's kind of like revoking an API key. This may not be an issue, because I'd also be encrypting this data, so I need to make some shared private key. When I revoke a team member, I'd have to change that private key, and then re-encrypt all of the team's data. So, it would kind of negate the need to do this, although you'd still want to revoke the ability to write files in some way.

**Comment by kantai** [comment]
> Can I de-authorize an auth token?

There's an expiration field, which can reduce the need for this, but it doesn't solve the problem entirely. Explicitly de-authorizing an auth token is hard to do right now, and probably requires an update to the auth token specification that is outside the scope of scoping.

**Comment by hstove** [comment]
I've tested this in a local app, with a code sample you can see at blockstack/blockstack.js#545 . It all worked perfectly for me :+1:

**Comment by jcnelson** [comment]
Cool!  So can this issue be closed?  Is this all merged in?

**Comment by kantai** [comment]
Merged to `develop`, but it hasn't been released to `master` yet -- I was planning on holding off until other infrastructure changes aren't in flux.

**Comment by larrysalibra** [comment]
> There's an expiration field, which can reduce the need for this, but it doesn't solve the problem entirely. Explicitly de-authorizing an auth token is hard to do right now, and probably requires an update to the auth token specification that is outside the scope of scoping.

The lack of way to de-authorize seems like a huge problem.

**Comment by larrysalibra** [comment]
@kantai wrote
> But here's roughly what I'd like to see in a scoped auth token:

>> pathsAllowed: [ string ],
maxFileSize: number
What were you thinking for the pathsAllowed? Supporting wildcards, or just explicit paths?

Did the file size restriction get in here?

From my brief reading of the merged pull request, this feature gives a 3rd party ability to write an unlimited amount of arbitrarily sized data to a prefixed area of user's storage.

**Comment by jcnelson** [comment]
> There's an expiration field, which can reduce the need for this, but it doesn't solve the problem entirely. Explicitly de-authorizing an auth token is hard to do right now, and probably requires an update to the auth token specification that is outside the scope of scoping.

> The lack of way to de-authorize seems like a huge problem.

Now that we have an administrative side-car, we have the opportunity to address this via the "bumping" strategy @kantai described elsewhere.  The Gaia config file can have a "minimum auth token nonces" object that maps addresses to integers.  A scoped auth token needs to have a nonce equal to or greater than the nonce for that address in the config file.  The admin side-car can increment the nonce on a per-app basis, thereby invalidating all outstanding auth tokens.

**Comment by hstove** [comment]
> Now that we have an administrative side-car, we have the opportunity to address this via the "bumping" strategy @kantai described elsewhere. The Gaia config file can have a "minimum auth token nonces" object that maps addresses to integers. A scoped auth token needs to have a nonce equal to or greater than the nonce for that address in the config file. The admin side-car can increment the nonce on a per-app basis, thereby invalidating all outstanding auth tokens.

I think I'm not following something here. The person who needs to 'expire' the token is the issuer, not a Gaia admin.

**Comment by kantai** [comment]
I talk about a revocation scheme in this issue:

https://github.com/blockstack/gaia/issues/150#issuecomment-442859882
